py/mpconfig.h

```diff
@@ -1548,4 +1553,14 @@ typedef double mp_float_t;
 # define MP_WARN_CAT(x) (NULL)
 #endif
 
+// Feature dependency check.
+#if MICROPY_PY_SYS_SETTRACE
+#if !MICROPY_PERSISTENT_CODE_SAVE
+#error "MICROPY_PY_SYS_SETTRACE requires MICROPY_PERSISTENT_CODE_SAVE to be enabled!"
+#endif
+#if MICROPY_COMP_CONST
+#error "MICROPY_PY_SYS_SETTRACE requires MICROPY_COMP_CONST to be disabled!"
```

Comment:
Why does it require COMP_CONST to be disabled?  Most ports have this feature enabled so it'd be good to make it work with settrace (not saying that it should be done now, just wondering what the problem is having them both enabled).

Reply:
It works with COMP_CONST enabled but the results are not precise -- compared to CPython. The uPy compiler optimisation enabled by COMP_CONST makes some lines _invisible_ eg.
```python
1. from micropython import const
2. _CNT02 = const(123)
```
Line 2 in the above example would be optimised out which is bad for the code coverage tool built upon this feature.

Changing the `error` **requirement** of `MICROPY_COMP_CONST` to `warning` **suggestion** would be more correct way to go.

Something like:
```c
#warning "MICROPY_PY_SYS_SETTRACE suggests MICROPY_COMP_CONST to be disabled for accurate traces.
```

Reply:
I see.  Maybe when settrace is enabled the emitter can emit a dummy operation in place here, like LOAD_CONST_NONE + POP_TOP.

Reply:
> Maybe when settrace is enabled the emitter can emit a dummy operation in place here, like LOAD_CONST_NONE + POP_TOP.

I'm afraid that would create confusion when the OPCODE tracing will be supported.

Another way (I didn't test it) is to change all `#if MICROPY_COMP_CONST` to `#if MICROPY_COMP_CONST && ! MICROPY_PY_SYS_SETTRACE` in _parse.c_.

If any auto-magic :dizzy: is desired I think this would be the one that would make anybody less angry about unexpected behaviour :smile: Anyway any auto-magic in this case creates coupling and doesn't keep the logic of enabled features explicit.

Reply:
Ok, let's leave it as you have it.  The point is (as you mention above) that to get precise results the "x=const(123)" feature must be turned off.